File: scripts/_recurrence.py
"""
Occurrence-date generation for recurring task series (#468).

The whole recurrence model lives here so the spawner stays a thin I/O shell and the rules are
testable without a database. Three freqs, an interval, and an optional weekday set — deliberately
not RFC 5545. See supabase/migrations/20260824000000_task_series.sql for why.

Import as: from _recurrence import occurrences_between, describe_cadence
"""
from __future__ import annotations

from calendar import monthrange
from datetime import date, timedelta

FREQS = ("daily", "weekly", "monthly")

# 0=Sun … 6=Sat — matches Postgres extract(dow) and JS Date.getDay(). Python's date.weekday() is
# Mon=0, so every conversion goes through this helper rather than an inline +1 % 7 scattered around.
_WEEKDAY_NAMES = ("Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday")


def dow(d: date) -> int:
    """Day of week as 0=Sun … 6=Sat."""
    return (d.weekday() + 1) % 7


def _add_months(d: date, months: int) -> date:
    """Shift by whole months, clamping to the end of the target month.

    Jan 31 + 1 month is Feb 28 (or 29), not Mar 3. Rolling over into the next month would make a
    monthly series drift forward a few days every short month until it no longer means anything.
    """
    total = (d.year * 12 + (d.month - 1)) + months
    year, month = divmod(total, 12)
    month += 1
    return date(year, month, min(d.day, monthrange(year, month)[1]))


def occurrences_between(
    *,
    freq: str,
    interval: int,
    byweekday: list[int] | None,
    starts_on: date,
    ends_on: date | None,
    window_start: date,
    window_end: date,
) -> list[date]:
    """Occurrence dates for a series that fall within [window_start, window_end].

    Dates are always derived from `starts_on`, never from the previous occurrence or from the
    completion of one — that is what stops a late-completed chore from dragging the whole series
    with it. Returns sorted, de-duplicated dates.
    """
    if freq not in FREQS:
        raise ValueError(f"unknown freq: {freq!r}")
    if interval < 1:
        raise ValueError(f"interval must be >= 1, got {interval}")

    # Clamp the requested window to the series' own lifetime before generating anything.
    lo = max(window_start, starts_on)
    hi = window_end if ends_on is None else min(window_end, ends_on)
    if lo > hi:
        return []

    out: list[date] = []

    if freq == "daily":
        # Snap forward to the first on-cadence day at or after `lo`, then step by interval.
        delta = (lo - starts_on).days
        step_index = -(-delta // interval)  # ceil division
        cur = starts_on + timedelta(days=step_index * interval)
        while cur <= hi:
            out.append(cur)
            cur += timedelta(days=interval)

    elif freq == "weekly":
        # An empty/absent byweekday means "the weekday starts_on falls on" — a weekly series always
        # has at least one day, and defaulting keeps a half-filled form from producing nothing.
        days = sorted(set(byweekday)) if byweekday else [dow(starts_on)]
        # Weeks are counted from the Sunday of the start week so that `interval` means "every N
        # weeks" consistently, regardless of which weekday the series started on.
        week0 = starts_on - timedelta(days=dow(starts_on))
        cur_week = week0
        # Skip whole weeks cheaply rather than walking day by day from starts_on.
        if lo > cur_week:
            weeks_elapsed = (lo - cur_week).days // 7
            cur_week += timedelta(days=(weeks_elapsed - (weeks_elapsed % interval)) * 7)
        while cur_week <= hi:
            for wd in days:
                d = cur_week + timedelta(days=wd)
                # `>= starts_on` matters on the first week: a Mon/Thu series starting Wednesday must
                # not emit that week's Monday, which is before the series existed.
                if lo <= d <= hi and d >= starts_on:
                    out.append(d)
            cur_week += timedelta(days=7 * interval)

    else:  # monthly
        # Walk month steps from starts_on. Bounded by `hi`, so the loop is short.
        n = 0
        while True:
            d = _add_months(starts_on, n * interval)
            if d > hi:
                break
            if d >= lo:
                out.append(d)
            n += 1

    return sorted(set(out))


def describe_cadence(
    *, freq: str, interval: int, byweekday: list[int] | None, ends_on: date | None = None
) -> str:
    """Human phrasing for a series, e.g. 'every Sunday', 'every 3 days', 'monthly until 2026-12-31'.

    Kept in step with cadenceLabel() in web/src/lib/tasks/recurrence.ts — the notification body and
    the task row should not describe the same series differently.
    """
    if freq == "daily":
        base = "every day" if interval == 1 else f"every {interval} days"
    elif freq == "weekly":
        days = sorted(set(byweekday)) if byweekday else []
        if interval == 1 and len(days) == 1:
            base = f"every {_WEEKDAY_NAMES[days[0]]}"
        elif interval == 1 and days:
            base = "every " + ", ".join(_WEEKDAY_NAMES[d][:3] for d in days)
        elif interval == 1:
            base = "every week"
        else:
            base = f"every {interval} weeks"
            if days:
                base += " on " + ", ".join(_WEEKDAY_NAMES[d][:3] for d in days)
    else:
        base = "every month" if interval == 1 else f"every {interval} months"

    return f"{base} until {ends_on.isoformat()}" if ends_on else base

File: web/src/lib/tasks/recurrence.ts
/**
 * Recurring-task series helpers (#468).
 *
 * The occurrence GENERATION lives in Python (scripts/_recurrence.py) because the spawner is a cron
 * job — this side only ever has to describe a series and validate a form, never materialize dates.
 * Keep `cadenceLabel` in step with `describe_cadence()` there: the ntfy body and the task row
 * should not describe the same series differently.
 */

export const FREQS = ["daily", "weekly", "monthly"] as const;
export type Freq = (typeof FREQS)[number];

/** 0=Sun … 6=Sat — matches Postgres extract(dow) and Date.getDay(), so neither side remaps. */
export const WEEKDAY_NAMES = [
  "Sunday",
  "Monday",
  "Tuesday",
  "Wednesday",
  "Thursday",
  "Friday",
  "Saturday",
] as const;

export const WEEKDAY_SHORT = ["Su", "Mo", "Tu", "We", "Th", "Fr", "Sa"] as const;

export interface CadenceInput {
  freq: Freq;
  interval: number;
  byweekday?: number[] | null;
  ends_on?: string | null;
}

/** Human phrasing, e.g. "every Sunday", "every 3 days", "every Mon, Thu". */
export function cadenceLabel({ freq, interval, byweekday }: CadenceInput): string {
  const n = Math.max(1, Math.trunc(interval || 1));
  if (freq === "daily") return n === 1 ? "every day" : `every ${n} days`;

  if (freq === "weekly") {
    const days = [...new Set(byweekday ?? [])].sort((a, b) => a - b);
    if (n === 1 && days.length === 1) return `every ${WEEKDAY_NAMES[days[0]]}`;
    if (n === 1 && days.length > 1)
      return `every ${days.map((d) => WEEKDAY_NAMES[d].slice(0, 3)).join(", ")}`;
    if (n === 1) return "every week";
    const base = `every ${n} weeks`;
    return days.length
      ? `${base} on ${days.map((d) => WEEKDAY_NAMES[d].slice(0, 3)).join(", ")}`
      : base;
  }

  return n === 1 ? "every month" : `every ${n} months`;
}

/** Cadence plus its end date, e.g. "every Sunday until 2026-12-31". */
export function cadenceLabelWithEnd(input: CadenceInput): string {
  const base = cadenceLabel(input);
  return input.ends_on ? `${base} until ${input.ends_on}` : base;
}

/** Whole days from today to `ends_on`, or null for an open-ended series. Negative = already past. */
export function daysUntilEnd(ends_on: string | null | undefined, today: string): number | null {
  if (!ends_on) return null;
  const a = new Date(`${ends_on}T00:00:00`).getTime();
  const b = new Date(`${today}T00:00:00`).getTime();
  if (Number.isNaN(a) || Number.isNaN(b)) return null;
  return Math.round((a - b) / 86_400_000);
}

/**
 * Warning runway before expiry — mirrors warning_days() in scripts/check_series_expiring.py.
 * At least 14 days, more for a sparse cadence: a weekly series gives you fewer chances to notice
 * than a daily one.
 */
export function warningDays(freq: Freq, interval: number): number {
  const perUnit = freq === "monthly" ? 30 : freq === "weekly" ? 7 : 1;
  return Math.max(14, 3 * perUnit * Math.max(1, Math.trunc(interval || 1)));
}

/** True when a series is inside its warning runway and hasn't been dismissed. */
export function isExpiringSoon(
  series: {
    freq: Freq;
    interval: number;
    ends_on: string | null;
    expiry_dismissed_at: string | null;
  },
  today: string,
): boolean {
  if (series.expiry_dismissed_at) return false;
  const left = daysUntilEnd(series.ends_on, today);
  if (left === null || left < 0) return false;
  return left <= warningDays(series.freq, series.interval);
}

/** Add whole months, clamping to month end — Jan 31 + 1 month is Feb 28, not Mar 3. */
export function addMonths(dateStr: string, months: number): string {
  const [y, m, d] = dateStr.split("-").map(Number);
  const total = y * 12 + (m - 1) + months;
  const ny = Math.floor(total / 12);
  const nm = (total % 12) + 1;
  const lastDay = new Date(Date.UTC(ny, nm, 0)).getUTCDate();
  const nd = Math.min(d, lastDay);
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${ny}-${pad(nm)}-${pad(nd)}`;
}

export interface SeriesDraft {
  freq: Freq;
  interval: number;
  byweekday: number[];
  starts_on: string;
  ends_on: string | null;
}

/** Validate a series before it reaches the database. Returns an error string, or null if valid. */
export function validateSeries(d: SeriesDraft): string | null {
  if (!FREQS.includes(d.freq)) return `Unknown repeat frequency: ${d.freq}`;
  if (!Number.isInteger(d.interval) || d.interval < 1) return "Repeat interval must be 1 or more.";
  if (!/^\d{4}-\d{2}-\d{2}$/.test(d.starts_on)) return "Start date must be YYYY-MM-DD.";
  if (d.ends_on) {
    if (!/^\d{4}-\d{2}-\d{2}$/.test(d.ends_on)) return "End date must be YYYY-MM-DD.";
    if (d.ends_on < d.starts_on) return "End date must be on or after the start date.";
  }
  if (d.byweekday.some((n) => !Number.isInteger(n) || n < 0 || n > 6))
    return "Weekdays must be 0 (Sunday) through 6 (Saturday).";
  return null;
}
